src/librustc_mir/hair/pattern/mod.rs

```diff
@@ -403,9 +414,11 @@ impl<'a, 'tcx> PatternContext<'a, 'tcx> {
             PatKind::Lit(ref value) => self.lower_lit(value),
 
             PatKind::Range(ref lo_expr, ref hi_expr, end) => {
-                match (self.lower_lit(lo_expr), self.lower_lit(hi_expr)) {
-                    (PatternKind::Constant { value: lo },
-                     PatternKind::Constant { value: hi }) => {
+                match (
```

Comment:
Do we want to maintain the user type annotations in this case?

Reply:
Or is there some reason why user type annotations are always irrelevant in ranges? In that case, could you put it in a comment?

Reply:
So the answer is "only char and numeric types are allowed in range patterns", which means we don't care about user types. You could still have this in a comment.

Reply:
We still need them to check something like `Type<'a>::X` when we have `impl Type<'static> { const X: i32 = 0; ... }`

Reply:
So in that case, you need to move the `AscribeUserType` out.

Reply:
Added a comment.